plugins/hls-haddock-comments-plugin/src/Ide/Plugin/HaddockComments.hs

```diff
@@ -0,0 +1,154 @@
+    [(locDecl, src, x)] <- [(locDecl, l, x) | locDecl@(L l (fromDecl -> Just x)) <- locDecls, inRange range l],
+    annKeys <- collectKeys x,
+    not $ null annKeys,
+    and $ maybe False isFresh . flip Map.lookup anns <$> annKeys,
+    anns' <- foldr (Map.adjust updateAnn) anns annKeys,
+    Just range' <- calcRange src range,
+    result <- T.strip . T.pack $ exactPrint locDecl anns' =
+    Just (title, TextEdit range' result)
+  | otherwise = Nothing
+
+-----------------------------------------------------------------------------
+
+genForSig :: GenComments
+genForSig = GenComments {..}
+  where
+    title = "Generate signature comments"
+
+    fromDecl (SigD _ (TypeSig _ _ (HsWC _ (HsIB _ x)))) = Just x
+    fromDecl _ = Nothing
+    updateAnn x = x {annEntryDelta = DP (0, 1), annsDP = dp}
+
+    isFresh Ann {annsDP}
+      | null [() | (AnnComment _, _) <- annsDP] = True
+      | otherwise = False
+
+    collectKeys = keyFromTyVar 0
+
+    comment = mkComment "-- ^ " noSrcSpan
+    dp = [(AnnComment comment, DP (0, 1)), (G AnnRarrow, DP (1, 2))]
+
+genForRecord :: GenComments
+genForRecord = GenComments {..}
+  where
+    title = "Generate fields comments"
+
+    fromDecl (TyClD _ DataDecl {tcdDataDefn = HsDataDefn {dd_cons = cons}}) = Just [x | (L _ ConDeclH98 {con_args = x}) <- cons]
+    fromDecl _ = Nothing
+
+    updateAnn x = x {annEntryDelta = DP (1, -7), annPriorComments = [(comment, DP (1, -7))]}
+
+    isFresh Ann {annPriorComments} = null annPriorComments
+
+    collectKeys = keyFromCon
+
+    comment = mkComment "-- | " noSrcSpan
+
+-----------------------------------------------------------------------------
+
+toAction :: T.Text -> Uri -> TextEdit -> CodeAction
+toAction title uri edit = CodeAction {..}
+  where
+    _title = title
+    _kind = Just CodeActionQuickFix
+    _diagnostics = Nothing
+    _command = Nothing
+    _changes = Just $ HashMap.singleton uri $ List [edit]
+    _documentChanges = Nothing
+    _edit = Just WorkspaceEdit {..}
+
+calcRange :: SrcSpan -> Range -> Maybe Range
+calcRange src range
+  | inRange range src,
+    (RealSrcSpan span) <- src,
+    range' <- realSrcSpanToRange span =
+    Just range'
+  | otherwise = Nothing
+
+inRange :: Range -> SrcSpan -> Bool
+inRange range x = isInsideSrcSpan (_start range) x || isInsideSrcSpan (_end range) x
+
+-----------------------------------------------------------------------------
+
+keyFromTyVar :: Int -> LHsType GhcPs -> [AnnKey]
+keyFromTyVar dep c@(L _ (HsFunTy _ x y))
+  | dep < 1 = mkAnnKey c : keyFromTyVar dep x ++ keyFromTyVar dep y
+  | otherwise = []
+#if MIN_GHC_API_VERSION(8,10,0)
+keyFromTyVar dep (L _ (HsForAllTy _ _ _ x)) = keyFromTyVar dep x
+#else
+keyFromTyVar dep (L _ (HsForAllTy _ _ x)) = keyFromTyVar dep x
+#endif
+keyFromTyVar dep (L _ (HsQualTy _ _ x)) = keyFromTyVar dep x
+keyFromTyVar dep (L _ (HsParTy _ x)) = keyFromTyVar (succ dep) x
+keyFromTyVar dep (L _ (HsBangTy _ _ x)) = keyFromTyVar dep x
+keyFromTyVar _ _ = []
```

Comment:
I think this function should also do some work for other cases like, for example, `HsKindSig`.
It might be better to cover cases explicitly (so that we can notice the change when GHC introduce a new constructor.)

Reply:
Currently `HsType` has 23 constructors, but we only use five of them, so I think enumerating all cases would be a little bit distracting...